### ISSUE TITLE
URL slug front-end validation competes with back-end sanitization

### ISSUE THREAD

**Issue by anrghg** [WordPress/gutenberg#44444]
### Description

The Block Editor (as well as the Classic Editor) has an input field to edit the URL slug. As soon as this gains then loses focus, the slug is damaged in an uncontrollable way. While the `sanitize_title` filters are still applied in the back-end, the front-end preempts a number of simplifications, namely by removing diacritics and non-letter characters, defeating as much the purpose of the hook’s `$raw_title` argument.

The editors’ hasty busyness notwithstanding, the `sanitize_title` filters are still applied, yet they are overridden by what the front-end does. Conversely, the front-end does not do a full job, and the `sanitize_title` hook is still needed. The reason is, the front-end is uninformed, as it has no access to the `sanitize_title` hook.

So, the question is either:

Why do we have a `sanitize_title` hook, given the editor is empowered to edit the title at its end?

Or:

Why does the editor start dumbing down the title, given it is unable to complete, while disregarding the `sanitize_title` filters, and handing over a corrupted `$raw_title`?

The likely solution is to deactivate slug input field value front-end validation, and to delegate slug validation entirely to the back-end.

The remaining question is: Why do editors simplify slugs following their own rules **when that input field has focus** (but not when it never gets focus)? What considerations could fuel the decision to set up such a behavior?

Perhaps the fear users could consider the editor dumb when it does nothing by itself, waits for the back-end. Perhaps addressing a user expectation to see the editor proactively altering the slug **while it is on it.**

So, after applying the suggested fix, the editor will indeed look less “clever”. That would bring a need to educate users to consider WordPress as a whole, not to expect a unit battling another unit in an uncontrolled (bypassing the `sanitize_title` hook), disordered (ignoring the filters `sanitize_title` needs to apply) fight for reputation and “looking clever”.

I think such considerations about presumed user expectations should be ditched in favor of a consistent, regular, predictable WordPress, that well-advised users are expected… well, to expect.

### Step-by-step reproduction instructions

n/a

### Screenshots, screen recording, code snippet

n/a

### Environment info

n/a

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by anrghg** [comment]
@glendaviesnz Thanks for removing the Stale label.

@skorasaurus Front-end validation is considered state-of-the-art yet may be counter-productive and disturbing like here. This is not a matter of editors since the Classic editor is affected too. (I don’t use routinely the Classic editor, just for test purposes.)

**Comment by skorasaurus** [comment]
> @glendaviesnz Thanks for removing the Stale label.
> 
> @skorasaurus Front-end validation is considered state-of-the-art yet may be counter-productive and disturbing like here. This is not a matter of editors since the Classic editor is affected too. (I don’t use routinely the Classic editor, just for test purposes.)

I don't disagree; the needs technical feedback label is used to designate issues that need attention from developers so that they can give feedback on why things they are, from a technical perspective.

**Comment by anrghg** [comment]
> > @glendaviesnz Thanks for removing the Stale label.
> > @skorasaurus Front-end validation is considered state-of-the-art yet may be counter-productive and disturbing like here. This is not a matter of editors since the Classic editor is affected too. (I don’t use routinely the Classic editor, just for test purposes.)
> 
> I don't disagree; the needs technical feedback label is used to designate issues that need attention from developers so that they can give feedback on why things they are, from a technical perspective.

@skorasaurus Thanks for explaining. Providing such feedback may not be easy. I see a chance the feature was implemented only for the sake of making the editors more state-of-the-art, regardless of coming across the `sanitize_title` hook's purpose.

**Comment by tellthemachines** [comment]
@anrghg could you explain your use case a bit more? Do you need the accents preserved in $raw_title for any reason?

**Comment by anrghg** [comment]
@tellthemachines  Thank you for looking into this. The point is not so much in preserving accents, although these may also be valued depending on user preferences. Latin script has the misfortune of being often expected to be dumbed down by removing much more than just accents. One outstanding example is the removal of apostrophes. Apostrophes should rather be replaced with hyphens (https://plugins.trac.wordpress.org/browser/anrghg/tags/1.16.5/includes/modular.php#L1065).

But this is also a matter of symbols and emoji stripped alongside.

This slug input field is pretty unusable, since all the back-end fixes are swept away as soon as that field gets focus. As soon as the user makes the mistake of giving it focus.

The front-end slug validator is out of sync with `sanitize_title`, and the editor seems also to be lacking a hook that can perform in the front-end what we can do in the back-end thanks to `sanitize_title`.

**Comment by Marc-pi** [comment]
yep, i just noticed that emojis are kept in the slug / url (my users write sometimes a title with emojis in it). 
I doubt it's ok with RFCs :-)
I was looking for a way to sanatize them, per hook.

> Apostrophes should rather be replaced with hyphens" 

agree, we have to edit slugs for that purpose, would be fine if it can be automatic

> The point is not so much in preserving accents

agree, without them its better (at least in french). 
A hook would be a solution for those who want some customization

**Comment by tellthemachines** [comment]
At the very least it's a terrible experience to have an input change when the field loses focus. I'm going to label this as a bug.

**Comment by anrghg** [comment]
> I'm going to label this as a bug.

Thanks for granting the bug label.

> At the very least it's a terrible experience to have an input change when the field loses focus.

Not exactly. The slug’s backend-generated value changes when the field *gains* focus and the front-end disagrees.

**Comment by talldan** [comment]
This has quite a lot of technical feedback already. Just to break it down:

The difference in behavior boils down to frontend JS code differing in behavior to serverside PHP code.

When updating the 'Slug' input, a JavaScript function called `cleanForSlug` cleans the slug and strips emojis and some other characters. (https://github.com/WordPress/gutenberg/blob/trunk/packages/url/src/clean-for-slug.ts)

OTOH, if the post title contains an emoji, server side PHP code uses `sanitize_title` to turn it into a slug, and this seems to preserve emojis. (any complaints about that one need to be in [trac issues](https://core.trac.wordpress.org/), as the code doesn't live in the gutenberg repo).

I expect the reason `cleanForSlug` exists is to give more immediate feedback to users, but I can see that it's potentially an issue if it doesn't match `sanitize_title`.

It's probably never going to match given `sanitize_title` can be filtered. Perhaps the best thing is to remove `cleanForSlug` and depend on the server for correctness. 🤔

**Comment by anrghg** [comment]
> […] Perhaps the best thing is to remove `cleanForSlug` and depend on the server for correctness. 🤔

If `cleanForSlug` is removed cold turkey, the field may appear broken.

Unless `cleanForSlug` can be synced with `sanitize_title` and all its filters (seems unfeasible), it needs to be very strict, just to make sure anything it lets pass won’t be sanitized later on, since that can be disappointing the other way around, and even worse because the user didn’t get immediate feedback about it.

I’d strongly recommend adding a setting "Sanitize the slug field" in the edit section, on an opt-out basis, so that admins (or only top-level admins?) can switch `cleanForSlug` off. Then, when the field gets focus, JS may warn about back-end sanitization instead.

**Comment by mjaragon** [comment]
Noting one more example of this issue (still happening on WP 6.9.4):

When the title of a post has an emoji: 
* While the post is in draft, the slug does not include the emoji (editors might think the slug is sanitized)
* When the post is published, the emoji is added to the slug
* When clicking on the slug field, just by leaving the field, the emoji is removed from the slug.

See example screencast: 
https://www.loom.com/share/125e68c6f7964420ac7f366929fdcf96

**Comment by ramonjd** [comment]
I also took a quick look here, and it seems that folks are landing on the same root cause: `cleanForSlug` does a lot of guessing and doesn't align with the realities of `sanitize_title`. 

I agree with @talldan that we could stop using `cleanForSlug` where it's doing the most damage in the `onBlur` [sanitization](https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/post-url/index.js#L153). 

This would persist what the user typed. On save, the server's `sanitize_title` runs, the post entity refreshes, and the
input updates to the truth.

The trade off is between blur and save, the input shows a "dirty" slug.

A compromise could be to demote  `cleanForSlug`  to preview-only for new drafts, so it'd be still used as the last resort fallback. The server already publishes the correct value as `generated_slug` (see this [type](https://github.com/WordPress/gutenberg/blob/daf20d82b937e13d0f55e908008be8b3037c2d67/packages/core-data/src/entity-types/page.ts#L70)).

Here's a draft PR folks can test to make sure my assumptions are correct.

- https://github.com/WordPress/gutenberg/pull/78135/

Longer term, preview accuracy on unsaved drafts could be done via REST endpoint I guess, e.g., `/wp/v2/sanitize-title` which would be  filter +  locale aware. It'd also have some other potential uses:

- uniqueness or reserved word feedback
- headless editors

**Comment by ramonjd** [comment]
Reopening. Closed by accident. https://github.com/WordPress/gutenberg/pull/78135 didn't fix this issue, but just introduces a check for the server-generated slug from the post entity
